graph2tac/tfgnn/tasks.py

```diff
@@ -1206,6 +1231,37 @@ def _log_softmax_logits(
 
         return local_logits, global_logits
 
+    @staticmethod
+    def transpose_ragged_first_dim(
+        x: tf.RaggedTensor,  # [outer*inner, None(ragged), ...]
+        outer_dim: int,
+        inner_dim: int,
+    ) -> tf.RaggedTensor:  # [inner*outer, None(ragged), ...]
+        """
+        Convert a ragged tensor of shape `[outer*inner, None(ragged), ...]`
+        to one of shape `[inner*outer, None(ragged), ...]`
+        by permuting the inner values.
+        """
+        inner_dim = tf.cast(inner_dim, dtype=tf.int64)
+        outer_dim = tf.cast(outer_dim, dtype=tf.int64)
+
+        # calculate new row ids
+        rowids = x.value_rowids()  # [outer-inner-ragged]
+        batch1_rowids = rowids // inner_dim  # [outer-inner-ragged]
+        batch2_rowids = rowids % inner_dim   # [outer-inner-ragged]
+        new_rowids = batch2_rowids * outer_dim + batch1_rowids  # [outer-inner-ragged]
+        
+        # reorder the elements according to the new row ids
+        new_order = tf.argsort(new_rowids)  # [inner-outer-ragged]
+        new_rowids = tf.sort(new_rowids)  # [inner-outer-ragged]
```

Comment:
Couldn't the following lines be written just with some `tf.gather` / `tf.invert_permutation`? (not that I would understand fully the nuances of the code)

Reply:
I am not sure tf.invert_permute works here.  new_rowids isn't a permutation.  (Indeed there can be multiple batch elements with the same row ids.  I'm specifically using that tf.argsort maintains order when there is a tie.)  I guess instead of tf.sort I can could use gather on the sorted indices.  I don't know what is faster.  But also, this isn't really a slow operation in the code (and maybe I should just fix the need for having it in the first place).

Reply:
Oh, I need to set stable=true for that in the argsort.